core/dbt/graph/selector.py

```diff
@@ -329,16 +338,51 @@ def select_childrens_parents(self, selected: Set[str]) -> Set[str]:
         ancestors_for = self.select_children(selected) | selected
         return self.select_parents(ancestors_for) | ancestors_for
 
-    def select_children(self, selected: Set[str]) -> Set[str]:
+    @classmethod
+    def descendants(cls, graph, node, max_depth: int = None):
+        """Returns all nodes reachable from `node` in `graph`"""
+        if not graph.has_node(node):
+            raise nx.NetworkXError("The node %s is not in the graph." % node)
+        des = set(
+            n for n, d in
+            nx.single_source_shortest_path_length(G=graph,
+                                                  source=node,
+                                                  cutoff=max_depth).items()
+        )
+        return des - {node}
+
+    @classmethod
+    def ancestors(cls, graph, node, max_depth: int = None):
+        """Returns all nodes having a path to `node` in `graph`"""
+        if not graph.has_node(node):
+            raise nx.NetworkXError("The node %s is not in the graph." % node)
+        with nx.utils.reversed(graph):
+            anc = set(
+                n for n, d in
+                nx.single_source_shortest_path_length(G=graph,
+                                                      source=node,
+                                                      cutoff=max_depth).items()
+            )
+        return anc - {node}
```

Comment:
A couple things about this:
 - Why not just make these methods? Instead of `graph` here it'd be `self.graph` but you could remove the argument and replace `Graph.descendants(self.graph, ...)` with `self.descendants(...)`, which seems much simpler.
 - please use `{}` and `.format` (or f-strings!) for errors, not `%`-formatting.


- Can't those set() genexpr calls just be replaced with a single thing like this?:
```python
set(nx.single_source_shortest_path_length(G=Graph, source=node, cutoff=max_depth)
```

since `nx.single_source_shortest_path_length` returns a dict, set() on it will just get the keys, which is all you care about here.

Reply:
@beckjake Ohh. I should broadly let You know that it's only a dump of https://networkx.github.io/documentation/stable/_modules/networkx/algorithms/dag.html#ancestors just to give access to `cutoff`. It's was the reason for `format` and desired static method etc.

Reply:
I agree with all of the suggestions of course! I'm going to apply some refactors soon.

Reply:
Done